spec/models/email_unpublishing_processor_spec.rb

```diff
@@ -0,0 +1,234 @@
+      discard: nil,
+      retry: nil,
+      payload: document,
+    )
+  end
+
+  let(:processor) { EmailUnpublishingProcessor.new(logger) }
+  let(:mock_unpublishing_alert) { double("UnpublishingAlert", trigger: nil) }
+
+  let(:content_id) { SecureRandom.uuid }
+  let(:govuk_request_id) { SecureRandom.uuid }
+  let(:public_updated_at) { Time.now.iso8601 }
+
+  let(:published_in_error_payload) do
+    {
+      "document_type" => "gone",
+      "schema_name" => "gone",
+      "base_path" => "/foo",
+      "locale" => "en",
+      "publishing_app" => "whitehall",
+      "public_updated_at" => public_updated_at,
+      "details" => {
+        "explanation" => "Why the page was published in error",
+        "alternative_path" => alternative_path,
+      },
+      "content_id" => content_id,
+      "govuk_request_id" => govuk_request_id,
+      "payload_version" => 199,
+    }
+  end
+
+  let(:consolidated_payload) do
+    {
+      "document_type" => "redirect",
+      "schema_name" => "redirect",
+      "base_path" => "/foo",
+      "locale" => "en",
+      "publishing_app" => "whitehall",
+      "redirects" => [
+        {
+          "path" => "/government/publications/foo/html-attachment-foo",
+          "type" => "exact",
+          "destination" => "/government/publications/foobar",
+        },
+      ],
+      "public_updated_at" => public_updated_at,
+      "content_id" => content_id,
+      "govuk_request_id" => govuk_request_id,
+      "payload_version" => 199,
+    }
+  end
+
+  def unpublish_alert_was_called
+    expect(mock_unpublishing_alert).to have_received(:trigger)
+  end
+
+  def unpublish_alert_was_not_called
+    expect(mock_unpublishing_alert).not_to have_received(:trigger)
+  end
+
+  before :each do
+    allow(UnpublishingAlert).to receive(:new).and_return(mock_unpublishing_alert)
+  end
+
+  describe "#process" do
+    describe "in an unknown unpublishing scenario" do
+      let(:document) do
+        {
+          "document_type" => "woops",
+          "schema_name" => "woops",
+          "base_path" => "/foo",
+          "public_updated_at" => public_updated_at,
+          "content_id" => content_id,
+        }
+      end
+
+      it "does not trigger an email-api client call and logs a message" do
+        processor.process(message)
+
+        unpublish_alert_was_not_called
+        expect(logger).to have_received(:info).with(
+          "not triggering bulk unsubscription and alert for document with a unknown unpublishing scenario. Document type: #{document['document_type']}",
+        )
+      end
+    end
+
+    describe "following published_in_error_with_url event" do
+      let(:unpublishing_scenario) { :published_in_error_with_url }
+      let(:alternative_path) { "/foo/alternative" }
+      let(:document) { published_in_error_payload }
+
+      it "acknowledges and triggers an email-api client call for a correctly tagged document" do
+        processor.process(message)
+
+        unpublish_alert_was_called
```

Comment:
You should check what it was called with, to test that the correct unpublishing scenario is being passed.

Reply:
You can do

```ruby
expect(mock_unpublishing_alert).not_to have_received(:trigger).with(arguments, go, here)
```

Reply:
wouldn't that only work if `.trigger` took an argument?
https://github.com/alphagov/email-alert-service/pull/481/files#diff-662628447d5cb8aa42dc5c71945cbdc7f91ad85a5040ab272bb4276caa3f1aa2R11-R15

Reply:
Oh yes, of course. You'd have to do it in the `before`, something like:

```ruby
before :each do
  allow(UnpublishingAlert).to receive(:new).with(arguments, go, here).and_return(mock_unpublishing_alert)
end
```

(as an aside, isn't the `:each` there irrelevant? what else does `before` mean?)

The problem I'm trying to solve here is that I don't think we're actually testing that each test scenario triggers a different sort of email alert.  Each block defines a `unpublishing_scenario` variable, but it's unused.  So we could break that functionality, making them all return the same thing, and not notice.

Reply:
Right so the test needs to be something like.
Given a document, logger and unpublishing scenario.
We expect trigger to be called.
Whilst an instance of UnpublishingAlert has a certain state.

Let me quickly look up how the doubles work and what's accessible.

Unit tests on the `UnpublishingAlert` should cover a lot of this I think